internal/jobservice/repository/sql_job_service.go

```diff
@@ -0,0 +1,267 @@
+package repository
+
+import (
+	"database/sql"
+	"fmt"
+	"sync"
+	"time"
+
+	_ "modernc.org/sqlite"
+
+	"github.com/G-Research/armada/internal/jobservice/configuration"
+	js "github.com/G-Research/armada/pkg/api/jobservice"
+
+	log "github.com/sirupsen/logrus"
+)
+
+// Internal structure for storing in memory JobTables and Subscription JobSets
+// Locks are used for concurrent access of map
+type JobStatus struct {
+	jobMap        map[string]*JobTable
+	jobLock       sync.RWMutex
+	subscribeMap  map[string]*SubscribeTable
+	subscribeLock sync.RWMutex
+}
+
+func NewJobStatus(jobMap map[string]*JobTable, subscribeMap map[string]*SubscribeTable) *JobStatus {
+	return &JobStatus{jobMap: jobMap, subscribeMap: subscribeMap}
+}
+
+// SQLJobService for persisting to DB.
+type SQLJobService struct {
+	jobStatus        *JobStatus
+	jobServiceConfig *configuration.JobServiceConfiguration
+	db               *sql.DB
+}
+
+func NewSQLJobService(jobMap *JobStatus, config *configuration.JobServiceConfiguration, db *sql.DB) *SQLJobService {
+	return &SQLJobService{jobStatus: jobMap, jobServiceConfig: config, db: db}
+}
+
+type SubscribedTuple struct {
+	Queue  string
+	JobSet string
+}
+
+// Create a Table from a hard-coded schema.
+// Open to suggestions on how to make this better
+func (s *SQLJobService) CreateTable() {
+	_, err := s.db.Exec("DROP TABLE IF EXISTS jobservice")
+	if err != nil {
+		panic(err)
+	}
+	_, err = s.db.Exec(`
+CREATE TABLE jobservice (
+Queue TEXT,
+JobSetId TEXT,
+JobId TEXT,
+JobResponseState TEXT,
+JobResponseError TEXT,
+Timestamp INT
+)`)
+	if err != nil {
+		panic(err)
+	}
+}
+
+// Get the JobStatus given the jodId
+// If a job is not in the map, we return JOB_ID_NOT_FOUND
+// This should not be an error.
+func (s *SQLJobService) GetJobStatus(jobId string) (*js.JobServiceResponse, error) {
+	s.jobStatus.jobLock.RLock()
+	jobResponse, ok := s.jobStatus.jobMap[jobId]
+	s.jobStatus.jobLock.RUnlock()
+	if !ok {
+		sqlResponse, errSql := s.GetJobStatusSQL(jobId)
```

Comment:
AFAICT, the jobId is never sanitized against SQL injection at any point. We should ensure these inputs are well validated before feeding them to sqlite (and/or the SQL wrapper functions, depending on how you wanna structure it -- IMO it's best to do it closest to the DB call as possible).

Reply:
There must be some OSS sanitation library we can leverage for this purpose, right?

Reply:
I think there's something in the golang stdlib, actually, but I haven't looked in detail.

Reply:
Kevin and I talked about this; using parameterized queries like exist already ensure you can't do sql injection. It's so embedded in the stdlib you get it for borderline-free!

Reply:
SQL sanitization techniques are not needed here - the `GetJobStatusSQL()` func uses parameter binding - this robustly thwarts injection attempts at the driver level:
```
func (s *SQLJobService) GetJobStatusSQL(jobId string) (*js.JobServiceResponse, error) {
  row := s.db.QueryRow("SELECT JobResponseState, JobResponseError FROM jobservice WHERE JobId=?", jobId)
  ```
If there's any creation of complex SQL via string concatenation, then yeah, it gets risky very quickly.

Reply:
The classic sprintf() technique (it protects nothing) is warned against in https://tip.golang.org/doc/database/sql-injection

Reply:
Thanks all!